src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java

```diff
@@ -182,24 +183,48 @@ private Response buildOptionsResponse(Request request, ResponseBuilder builder)
   }
 
   private Response buildGetDirectoryResponse(Request request, ResponseBuilder builder) {
-    var contents =
+    var list =
         fileController.getDirectoryContents(request.uri.getPath())
                       .stream()
                       .sorted()
-                      .reduce((p1, p2) -> p1 + "," + p2)
-                      .orElse("");
-
-    if (contents.isEmpty()) {
+                      .map(p -> {
+                        // would be nice to append "/" to directories
+                        // dir items must not contain leading slash
+                        // or the <base> element won't work
+                        return "<a href=\"" + p + "\">" + p + "</a>";
+                      })
+                      .map(s -> "<li>" + s + "</li>")
+                      .collect(Collectors.joining("\n"));
+
+    if (list.isEmpty()) {
       return builder.withHeader("Content-Length", "0")
                     .withStatusCode(200)
                     .build();
     }
 
+    var body = String.join("\n", "<ul>", list, "</ul>");
+    var basePathBuilder = new StringBuilder(request.uri.toString());
+    if (basePathBuilder.charAt(0) != '/') {
+      basePathBuilder.insert('/', 0);
+    }
+    if (basePathBuilder.charAt(basePathBuilder.length() - 1) != '/') {
+      basePathBuilder.append('/');
+    }
+    var basePath = basePathBuilder.toString();
+    var contents =
+        String.join(
+            "\n",
+            "<head>",
+            "<base target=_self href=\"" + basePath + "\"/>",
+            "</head>",
+            "<body>",
+            body,
+            "</body>");
```

Comment:
> This HTML generation seems like a natural place for a seam...

Definitely something I'd like to refactor once I've confirmed that it's up & running.

> One potential design for this specific thing: a class that takes whatever base-directory config it needs in a constructor, with a method that takes a list of files and returns an HTML page as a string

The "feature requirements" here are pretty basic: generate an HTML page with a link for each file in the given directory. This could be (and maybe was) a couple-liner stream transform. Most of the other stuff is encapsulating HTML (head/body/etc.) so I could have a spot to put the `<base>` element. 

> I'll make another unit testing plug here too...

Fair, I probably should have considered using an "offline" HTML parser to test the HTML generation. However, I was less worried about getting that right than the actual navigation of the HTML in a browser (hence my recent addition of Selenium-driven tests). 

I'll see if I can pick any low-hanging fruit to speed up test execution.